conans/cli/command.py

```diff
@@ -0,0 +1,282 @@
+            # with the default as `values` for other arguments which
+            # share this destination.
+            parser.set_defaults(**{self.dest: None})
+
+        if isinstance(values, str):
+            dest.append(values)
+        elif values:
+            try:
+                dest.extend(values)
+            except ValueError:
+                dest.append(values)
+
+
+class OnceArgument(argparse.Action):
+    """Allows declaring a parameter that can have only one value, by default argparse takes the
+    latest declared and it's very confusing.
+    """
+
+    def __call__(self, parser, namespace, values, option_string=None):
+        if getattr(namespace, self.dest) is not None and self.default is None:
+            msg = '{o} can only be specified once'.format(o=option_string)
+            raise argparse.ArgumentError(None, msg)
+        setattr(namespace, self.dest, values)
+
+
+class SmartFormatter(argparse.HelpFormatter):
+
+    def _fill_text(self, text, width, indent):
+        text = textwrap.dedent(text)
+        return ''.join(indent + line for line in text.splitlines(True))
+
+
+class ConanCommand(object):
+    def __init__(self, method, group=None):
+        self._group = group if group is not None else "Misc commands"
+        self._name = method.__name__.replace("_", "-")
+        self._method = method
+        self._doc = method.__doc__
+        self._parser = argparse.ArgumentParser(description=self._doc,
+                                               prog="conan {}".format(self._name),
+                                               formatter_class=SmartFormatter)
+
+    @property
+    def group(self):
+        return self._group
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def method(self):
+        return self._method
+
+    @property
+    def doc(self):
+        return self._doc
+
+    @property
+    def parser(self):
+        return self._parser
+
+
+def conan_command(group=None):
+    def decorator(f):
+        cmd = ConanCommand(f, group)
+        return cmd
+
+    return decorator
+
+
+class Command(object):
+    """A single command of the conan application, with all the first level commands. Manages the
+    parsing of parameters and delegates functionality to the conan python api. It can also show the
+    help of the tool.
+    """
+
+    def __init__(self, conan_api):
+        assert isinstance(conan_api, Conan)
+        self._conan = conan_api
+        self._out = conan_api.out
+        self._groups = {}
+        self._commands = None
+
+    def _add_command(self, import_path, method_name):
+        try:
+            command_wrapper = getattr(importlib.import_module(import_path), method_name)
+            if command_wrapper.doc:
+                self._commands[command_wrapper.name] = command_wrapper
+                self._groups.setdefault(command_wrapper.group, []).append(command_wrapper.name)
```

Comment:
together with the suggestion above

```suggestion
                self._groups[command_wrapper.group].append(command_wrapper.name)
```